### PR TITLE
Add support for NX action "learn" and "note"

### PR DESCRIPTION
**Pull request by wenyingd**
1. Add support for OVS extension actions "learn" and "note".
2. Rename Bundle error code prefix from "BFC" to "BEC".

Signed-off-by: wenyingd <wenyingd@vmware.com>

### CODE REVIEW

**Comment by wenyingd** [comment]
Hi @jojimt , could you help review this PR?

File: openflow13/nx_action.go [hunk 1]
```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
...
+	Header   *NXLearnSpecHeader
+	SrcField *NXLearnSpecField
+	DstField *NXLearnSpecField
+	SrcValue []byte
+}
+
+func (s *NXLearnSpec) Len() uint16 {
+	length := s.Header.Len()
+	if s.Header.src {
+		length += (s.Header.nBits + 15) / 16 * 2
```
Comment (jojimt) [inline openflow13/nx_action.go]: For better readability, please rewrite the expression as 2 * (s.Header.nBits + 15) / 16. Same comment for all such instances below.

File: openflow13/nx_action.go [hunk 2]
```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
...
+	if !s.Header.output {
+		length += 6
+	}
+	return length
+}
+
+func (s *NXLearnSpec) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, s.Len())
+	n := 0
+	b, err := s.Header.MarshalBinary()
```
Comment (jojimt) [inline openflow13/nx_action.go:1134]: Please check error.

File: openflow13/nx_action.go [hunk 3]
```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
...
+	copy(data[n:], b)
+	n += len(b)
+	var srcData []byte
+	var srcDataLength int
+	if s.Header.src {
+		srcDataLength = int((s.Header.nBits + 15) / 16 * 2)
+		srcData = s.SrcValue
+	} else {
+		srcData, err = s.SrcField.MarshalBinary()
```
Comment (jojimt) [inline openflow13/nx_action.go:1146]: Need to check and handle error.

File: openflow13/nx_action.go [hunk 4]
```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
...
+		srcDataLength = 6
+	}
+	copy(data[n:], srcData[:srcDataLength])
+	n += srcDataLength
+	if !s.Header.output {
+		var dstData []byte
+		dstData, err = s.DstField.MarshalBinary()
```
Comment (jojimt) [inline openflow13/nx_action.go:1156]: same here

File: openflow13/nx_action.go [hunk 5]
```diff
@@ -1003,3 +1005,329 @@ func NewNXActionDecTTLCntIDs(controllers uint16, ids ...uint16) *NXActionDecTTLC
...
+func (a *NXActionNote) Len() uint16 {
+	length := a.NXActionHeader.Len() + uint16(len(a.Note))
+	return (length + 7) / 8 * 8
+}
+
+func (a *NXActionNote) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, a.Len())
+	n := 0
+	a.Length = a.Len()
+	b, err := a.NXActionHeader.MarshalBinary()
```
Comment (jojimt) [inline openflow13/nx_action.go:1317]: check error

File: openflow13/nx_action.go (hunk 1 above)
Comment (wenyingd) [reply openflow13/nx_action.go]: Sure, updated.

File: openflow13/nx_action.go (hunk 2 above)
Comment (wenyingd) [reply openflow13/nx_action.go:1134]: Updated.

File: openflow13/nx_action.go (hunk 3 above)
Comment (wenyingd) [reply openflow13/nx_action.go:1146]: Updated.

File: openflow13/nx_action.go (hunk 4 above)
Comment (wenyingd) [reply openflow13/nx_action.go:1156]: Done.

File: openflow13/nx_action.go (hunk 5 above)
Comment (wenyingd) [reply openflow13/nx_action.go:1317]: Done.

**Comment by wenyingd** [comment]
Hi @jojimt , I have addressed your comments.

**Pull request merged**